package communication;

message Request {
  message Tuple {
    required string key = 1;
    optional string value = 2;
    optional uint64 timestamp = 3;
    optional uint32 num_address = 4;
  }
  required string type = 1;
  optional string respond_address = 2;
  repeated Tuple tuple = 3;
  optional string request_id = 4;
}

message Response {
  message Tuple {
    required uint32 err_number = 1;
    required string key = 2;
    optional string value = 3;
    optional uint64 timestamp = 4;
  }
  repeated Tuple tuple = 1;
  optional string response_id = 2;
}

message Key_Request {
  required string respond_address = 1;
  repeated string keys = 2;
  optional string request_id = 3;
}

message Key_Response {
  message Tuple {
    required string key = 1;
    repeated string addresses = 2;
  }
  repeated Tuple tuple = 1;
  optional string response_id = 2;
}

message Address {
  message Tuple {
    required string ip = 1;
  }
  required uint64 start_time = 1;
  repeated Tuple tuple = 2;
}
